### ISSUE TITLE
Scala 3 publishing plan

### ISSUE THREAD

**Issue by keynmol** [scalameta/scalameta#2485]
I did not find a tracking issue for this, but please point me to it if there is one.

For full transparency, I don't know if it's **necessary** to publish Scalameta for Scala 3. It would certainly be nice for end users (such as mdoc and **its** end users) not to propagate the 2.13 versions of certain libraries Scalameta uses.

Straight up publishing for Scala 3 is complicated by the heavy usage of Scala 2 macros, as such a very quick grep shows that at least this many macros/usages of macros will need to be re-written:

- [ ] /trees/shared/src/main/scala/scala/meta/internal/trees/AstInfo.scala
- [ ] /quasiquotes/shared/src/main/scala/scala/meta/internal/quasiquotes/ReificationMacros.scala
- [ ] /quasiquotes/shared/src/main/scala/scala/meta/internal/quasiquotes/ConversionMacros.scala
- [ ] /common/shared/src/main/scala/org/scalameta/adt/Liftables.scala
- [ ] /common/shared/src/main/scala/org/scalameta/adt/Adt.scala
- [ ] /common/shared/src/main/scala/org/scalameta/tests/TypecheckError.scala
- [ ] /common/shared/src/main/scala/org/scalameta/internal/FreeLocalFinder.scala
- [ ] /common/shared/src/main/scala/org/scalameta/internal/DebugFinder.scala
- [ ] /common/shared/src/main/scala/org/scalameta/internal/ImplTransformers.scala
- [ ] /common/shared/src/main/scala/org/scalameta/invariants/package.scala
- [ ] /common/shared/src/main/scala/org/scalameta/explore/Macros.scala
- [ ] /common/shared/src/main/scala/org/scalameta/package.scala
- [ ] /common/shared/src/main/scala/org/scalameta/data/data.scala
- [ ] /common/shared/src/main/scala/org/scalameta/data/Macros.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/trees/TyperMacros.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/trees/branch.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/trees/Liftables.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/trees/ast.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/trees/quasiquote.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/trees/root.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/trees/registry.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/prettyprinters/ShowMacros.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/transversers/transformer.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/transversers/traverser.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/transversers/transverser.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/classifiers/Macros.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/tokens/root.scala
- [ ] /common/shared/src/main/scala/scala/meta/internal/tokens/token.scala
- [ ] /common/shared/src/main/scala-2.13/org/scalameta/internal/MacroCompat.scala
- [ ] /dialects/shared/src/main/scala/scala/meta/dialects/package.scala
- [ ] /tokens/shared/src/main/scala/scala/meta/internal/tokens/TokenInfo.scala
- [ ] /src/test/scala-2.12/scala/meta/tests/metacp/MetacpErrorSuite.scala

And those macros are split into two major categories:

1. Macro annotations - which are currently not supported at all in Scala 3, for example: https://github.com/scalameta/scalameta/blob/main/scalameta/common/shared/src/main/scala/scala/meta/internal/trees/ast.scala#L11
    Faced with the same problem, Cats replaced their macros with Scalafix codegeneration: https://github.com/typelevel/simulacrum-scalafix without breaking any APIs.

2. Various other macros, whitebox and blackbox. Scala 3's macros are more restricted, but probably most things can be implemented, if may be using slightly different techniques.

### Fastparse 

Scalameta depends on Fastparse but not for the main parser, rather the auxiliary ones - XML and Scaladoc, e.g. https://github.com/scalameta/scalameta/blob/fb8bee548061d75397b5fea9fb5a3d3e1aabecd0/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScaladocParser.scala

Fastparse is a heavily macro-based project, and while some preparation work has started for Scala 3 port: https://github.com/com-lihaoyi/fastparse/pull/252 I'm not sure it will be easy or will progress quickly.

Fast parser options are thin on the ground for Scala 3 atm, cats-parse being the only one published and comparable in performance (afaik). Perhaps a hand-rolled parser will be easier or a macro-less version of Fastparse for Scala 3...

--- 

**Consideration**: Scalameta is an active project, powering several very important tools in the ecosystem, and the macro work above is a multi-week/month invasive undertaking.

To make it possible, some preliminary work needs to happen:

- [ ] Inline sourcecode dependency in the https://github.com/scalameta/fastparse fork.
   - Should be inlined in `scala.meta.internal.sourcecode` package and dependency removed from the build
   - https://github.com/scalameta/fastparse/pull/3
- [ ] Inline geny dependency in the https://github.com/scalameta/fastparse fork.
   - Should be inlined in `scala.meta.internal.geny` package and dependency removed from the build
   - https://github.com/scalameta/fastparse/pull/3
- [ ] Modify build to add Scala 3 versions where it is necessary (i.e. ignore Native, ignore Scalac plugin, etc.)
- [ ] Move macro definition and usage files into `scala-2` folders in the `common` module
        - This allows the Scala 2 version to compile and operate without changes 
        - The work on porting the macros can continue in parallel in separate sources, `scala-3` folder
- [ ] Keep as much code as possible in the `src/main/scala`, if it can be cross-compiled (I've done this exercise, there's not that many files, sadly)
- [ ] Disable full cross-compilation on CI, set Scala 3 sources compilation/testing as a separate workflow which doesn't fail the build 
- [ ] Gradually build out the meta machinery needed to compile Scalameta's code.

**Comment by olafurpg** [comment]
This looks like a great plan! Thank you for initiating this discussion. 

What 2.13 dependencies cause most problems for downstream users? I’m curious if we can replace them somehow so scalameta_2.13 will be dependency free or only depend on Scala 3 libraries (using the 2.13 tasty reader).

**Comment by keynmol** [comment]
```
└─ org.scalameta:scalameta_2.13:latest.stable -> 4.4.28 (possible incompatibility)
   ├─ org.scala-lang:scala-library:2.13.6
   ├─ org.scala-lang:scalap:2.13.6
   │  └─ org.scala-lang:scala-compiler:2.13.6
   │     ├─ net.java.dev.jna:jna:5.3.1
   │     ├─ org.jline:jline:3.19.0
   │     ├─ org.scala-lang:scala-library:2.13.6
   │     └─ org.scala-lang:scala-reflect:2.13.6
   │        └─ org.scala-lang:scala-library:2.13.6
   └─ org.scalameta:parsers_2.13:4.4.28
      ├─ org.scala-lang:scala-library:2.13.6
      └─ org.scalameta:trees_2.13:4.4.28
         ├─ org.scala-lang:scala-library:2.13.6
         ├─ org.scalameta:common_2.13:4.4.28
         │  ├─ com.lihaoyi:sourcecode_2.13:0.2.7
         │  ├─ com.thesamet.scalapb:scalapb-runtime_2.13:0.11.4
         │  │  ├─ com.google.protobuf:protobuf-java:3.15.8
         │  │  ├─ com.thesamet.scalapb:lenses_2.13:0.11.4
         │  │  │  ├─ org.scala-lang:scala-library:2.13.6
         │  │  │  └─ org.scala-lang.modules:scala-collection-compat_2.13:2.4.4
         │  │  │     └─ org.scala-lang:scala-library:2.13.5 -> 2.13.6
         │  │  ├─ org.scala-lang:scala-library:2.13.6
         │  │  └─ org.scala-lang.modules:scala-collection-compat_2.13:2.4.4
         │  │     └─ org.scala-lang:scala-library:2.13.5 -> 2.13.6
         │  └─ org.scala-lang:scala-library:2.13.6
         └─ org.scalameta:fastparse-v2_2.13:2.3.1
            ├─ com.lihaoyi:geny_2.13:0.6.5
            └─ com.lihaoyi:sourcecode_2.13:0.2.3 -> 0.2.7
```

The common annoyances are:

1. `geny` (coming from shadowed fastparse)
2. `sourcecode` (^ same)
3. `scala-collection-compat`
4. scala protobuf stuff

**Comment by keynmol** [comment]
Given that fastparse is already shadowed, we can do the same with geny and sourcecode

**Comment by olafurpg** [comment]
Is it possible to use the 2.13 tasty reader to upgrade those dependencies to Scala 3?

**Comment by keynmol** [comment]
So you mean having fastparse_2.13 depend on geny_3 and sourcecode_3 instead?

I think this will lead to opposite problem in 2.13 artifacts downstream (i.e. mdoc_2.13), as they will now have _3 artifacts resolved and SBT will barf again.

Is this what you meant?

**Comment by olafurpg** [comment]
You are right, that will cause problems for 2.13 artifacts. Can we inline those dependencies instead (excluding protobuf) to make them disappear? Both geny and sourcecode are tiny, and I suspect we're not using scala-collection-compat in many places.

**Comment by olafurpg** [comment]
It would of course be ideal if we can build Scalameta against Scala 3, but the checklist above shows that it will require a lot of effort. There shouldn't be any technical problems for Scala 3 users to consume scalameta_2.13 if it doesn't bring transitive 2.13 dependencies.

**Comment by keynmol** [comment]
Yep, I'm all for inlining at first, to remedy the current pain points.

I'm still eager to at least start on some of the macros, but may be in a separate sbt module. Will need a bit of rejigging to have at least some tests run against those, but I want to minimise the churn for the contributors to the main parts of scalameta.

**Comment by keynmol** [comment]
Modified the plan to inline both sourcecode and geny first, as it's best risk/reward move at this stage.

**Comment by tgodzik** [comment]
> @keynmol what's the status (beyond the comments in here) on this effort? I'm sure it's a big effort but it's forcing me into 2.13.10 vs 3.x since I need quasiquote support.

We discussed it recently during the Scala tooling summit and this is non-trivial migration. We might at some point pursue a simpler macro system for Scala 3, but we haven't decided on anything yet.

**Comment by smarter** [comment]
Here are my notes from the tooling summit, note that this is focused on porting all of scalameta which is why it mentions the core macro annotation, but we might instead prioritize porting the quasiquotes macros to make scalameta more usable from Scala 3, we didn't investigate those during the summit.

During our in-person discussion we mostly focused on the `@ast` macro annotation since this is the core macro annotation used to define the [scalameta ast](https://github.com/scalameta/scalameta/blob/main/scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala).
- The annotation is defined in https://github.com/scalameta/scalameta/blob/main/scalameta/common/shared/src/main/scala/scala/meta/internal/trees/ast.scala, it is rather long but there's comments along the way describing each step of the form `// step X: ...`
- https://github.com/scalameta/scalameta/blob/main/scalameta/common/shared/src/main/scala/scala/meta/internal/trees/ast.md shows an example expansion of the ast, although it isn't completely up-to-date anymore.
- Several of the steps in the macro annotation do not require adding new public members and could be implemented using the new experimental Scala 3 macro annotation support.
- ... but there's also quite a lot of public definitions added. These could be expanded using a scalafix rule.
- A lot of the complexity of the macro is related to its efficient lazy copying mechanism (see https://github.com/scalameta/scalameta/blob/8b47a881941602be241e5f6d6cf653703a850f09/scalameta/common/shared/src/main/scala/scala/meta/internal/trees/ast.scala#L143-L152). An initial version of a Scala 3 implementation could perhaps skip those.

**Comment by nafg** [comment]
Any guess when this might happen?

**Comment by tgodzik** [comment]
Not currently no, but you can use the Scalameta 2.13.x version in Scala 3 currently.

**Comment by henricook** [comment]
Would you mind giving me a dummy's explanation of how to use Scalameta 2.13.x in my Scala 3 project please?

**Comment by mkurz** [comment]
@henricook 
```sbt
libraryDependencies += ("org.scalameta" %% "scalameta" % "4.8.14").cross(CrossVersion.for3Use2_13),
```
We use parser like this in twirl: https://github.com/playframework/twirl/blob/382474d3ebc3761b37cb1af7138926fcb4364159/build.sbt#L119

**Comment by nafg** [comment]
Got bitten by zio-http-gen for scala 3 which uses scalameta-parsers pulling in sourcecode for 2.13, is there any hope of this getting fixed?

**Comment by tgodzik** [comment]
Yes, I just need to get the final part in, which means rewriting XMLParser we have that uses sourcecode. 

The WiP is here https://github.com/scalameta/scalameta/pull/3939

If anyone feels up to the challenge the only real thing left to do is to use the same xml parsing as we do for Scala 2 instead of using fastparse as it's almost impossible to make it sensibly work now. This can be done separately from the current WiP PR

**Comment by tgodzik** [comment]
My current approach is to switch to an xml scanner, which will not verify the actual XML, which is a bit of a regression, but on the other hand XML is a can of worms really :sweat_smile:

**Comment by tgodzik** [comment]
Good news! We've just merged the PR that added support for Scala 3 for all modules aside from Scala 3.

We have quasiquotes incoming and we'll be left with traversers and maybe changing trees modules at some point as well.

**Comment by tgodzik** [comment]
Closing this and opening other issues with smaller scope such as https://github.com/scalameta/scalameta/issues/4145

**Comment by tgodzik** [comment]
And https://github.com/scalameta/scalameta/issues/4146

**Comment by bjaglin** [comment]
Great work!

What about `org.scalameta:common` though? https://github.com/scalameta/scalameta/blob/f029adb3e0e3422fdf17ce84c5072a8888801f9d/build.sbt#L215

It brings com.lihaoyi:sourcecode which won't play nice if the client has it as well.

**Comment by tgodzik** [comment]
> What about org.scalameta:common though?

That's really intertwined with trees module, so both will need to be updated

> It brings com.lihaoyi:sourcecode which won't play nice if the client has it as well

I think we tried shading it, but that broke something and got reverted https://github.com/scalameta/scalameta/pull/3483

Not sure why it was, but it's also possible to just inline the parts we need.
